Integrate Link Checker API

emmabeynon opened the pull request:
Paired with @thomasleese 

[Trello Card](https://trello.com/c/QljohdgW/968-complete-local-links-manager-integration)

emmabeynon committed:
Add link_errors and link_warnings to Links

emmabeynon committed:
Add Webhooks controller

emmabeynon committed:
Add LinkStatusRequester

emmabeynon committed:
Add tests for LinkCheckerAPI requests

emmabeynon committed:
Modify scopes for Links

Links which have warnings will now appear under the 'Good links' tab in this
app, so we have changed the `:good_links` and `:currently_broken` scopes on the
Link model to reflect the new definitions of good links and broken links.

emmabeynon committed:
Update LinkStatusUpdater

This has been modified to work with the LinkStatusRequester, and also now
updated LocalAuthority homepage URLS.

emmabeynon committed:
Add link_errors and link_warnings to LocalAuthority

emmabeynon committed:
Remove LinkChecker

We are now using Link-Checker-API to check links in Local Links Manager, so we
no longer need a LinkChecker.

emmabeynon committed:
Remove HomepageStatusUpdater

Homepage statuses are now being updated in the LinkStatusUpdater, so we no
longer need the HomepageStatusUpdater.

thomasleese committed:
Update to the latest version of GDS API Adapters

thomasleese committed:
Use GDS API Adapters to talk to the Link Checker API

thomasleese committed:
Update status to use new value from Link Checker API

kevindew commented on spec/factories/links.rb:
```diff
@@ -4,6 +4,8 @@
     sequence(:url) { |n| "http://www.example.com/#{n}" }
     status nil
+    link_errors ""
+    link_warnings ""
```
Should these be empty hashes?

kevindew commented on lib/local-links-manager/check_links/link_status_requester.rb:
```diff
@@ -0,0 +1,31 @@
+  module CheckLinks
+    class LinkStatusRequester
+      #FIX-ME make this load from routes
+      CALLBACK_URI = Plek.find('local-links-manager') + '/link-check-callback'
```
it'd probably make most sense to inject this url in via the initializer. if you set an `as: :link_checker_callback` or similar you can use `app.link_checker_callback_url` - I guess if this is run from a rake context you'll probably have to configure a default hostname.

kevindew commented on spec/lib/local-links-manager/check_links/link_status_requester_spec.rb:
```diff
@@ -0,0 +1,49 @@
+require 'rails_helper'
+require 'local-links-manager/check_links/link_status_requester'
```
is there an autoloading issue?

thomasleese replied:
Local links manager doesn't seem to have autoloading. I would say out of the scope of this PR?

kevindew commented on spec/lib/local-links-manager/check_links/link_status_requester_spec.rb:
```diff
@@ -0,0 +1,49 @@
+    let(:local_authority_1) { FactoryGirl.create(:local_authority) }
+    let(:local_authority_2) { FactoryGirl.create(:local_authority) }
+    let!(:link_1) { FactoryGirl.create(:link, local_authority: local_authority_1, url: 'http://www.example.com') }
+    let!(:link_2) { FactoryGirl.create(:link, local_authority: local_authority_2, url: 'http://www.example.com/exampl.html') }
```
typo or intentional?

thomasleese replied:
Typo

emmabeynon replied:
Now fixed ⚡️

kevindew commented on spec/lib/local-links-manager/check_links/link_status_requester_spec.rb:
```diff
@@ -0,0 +1,49 @@
+      link_status_requester.call
+      expect(stub).to have_been_requested
+    end
+  end
```
I'm a bit confused here - which is probably my brain being the problem. Are we testing here that it only tests the homepage url for a disabled service link and not other links?

If so, should we be setting up a local auth and some links for it to not have?

emmabeynon replied:
The link factory has a local authority association, so a local authority (and also homepage) is automatically created for the link :) However, we probably should test that the disabled link is not included in the batch, so I'll add that in.

kevindew commented on spec/requests/link_checker_api_spec.rb:
```diff
@@ -0,0 +1,58 @@
+
+  before do
+    @time = Timecop.freeze('2016-06-21 09:26:56 +0100')
+  end
```
I think this would make more sense within the context

kevindew commented on spec/requests/link_checker_api_spec.rb:
```diff
@@ -0,0 +1,58 @@
+          ]
+        }
+
+        post '/link-check-callback', params: payload.to_json
```
This probably also makes sense in the before block which would make this `it` block a lot shorter

kevindew commented on spec/requests/link_checker_api_spec.rb:
```diff
@@ -0,0 +1,58 @@
+require 'rails_helper'
```
We've got a bit of mix of single and double quotes on the strings in this file

thomasleese replied:
Local links manager doesn't seem to be as strict on the quotes as `publishing-api`...

kevindew commented on app/models/link.rb:
```diff
@@ -18,10 +18,8 @@ class Link < ApplicationRecord
-  scope :good_links, -> { where(status: HTTP_OK_STATUS_CODE) }
-  scope :currently_broken, -> { where.not(status: HTTP_OK_STATUS_CODE) }
+  scope :good_links, -> { where(status: "ok") }
+  scope :currently_broken, -> { where.not(status: "ok") }
```
I'd think we only consider links with a status of broken as `currently_broken` Would have thought caution would be a separate grouping. Or if it'll be just one grouping it could be "For Review" rather than "Broken"? What do you think? Maybe Mia will have thoughts?

kevindew commented on app/models/link.rb:
```diff
@@ -60,7 +58,12 @@ def link_with_matching_url
+      @_link ||= link.first
+    else
+      @_link
+    end
```
I'm a bit confused about this (both versions of it) can we chat about it?

emmabeynon replied:
This had to be changed as a result of `link_errors` and `link_warnings` storing JSON.  After this change was made, the `distinct` method stopped working due to a Postgres error `could not identify an equality operator for type json`.  After doing some research I found this could be fixed by changing the attributes' type to jsonb, however this requires a newer version of Postgres, so I used `group_by` as a workaround to avoid having to upgrade Postgres. Not very pretty, but it does the job!

kevindew commented on lib/local-links-manager/check_links/link_checker.rb:
```diff
@@ -1,46 +0,0 @@
-module LocalLinksManager
```
Yay all the code going :tada:

kevindew reviewed (commented):
Looking really good 👍  - a bunch of comments included, the outdated ones still seem to apply so left them in there

emmabeynon committed:
Skip CSRF check for Link Checker API

We also require the LinkStatusUpdater file which is needed in order to update
links with the checks received from Link Checker API.

emmabeynon committed:
Change presentation of link statuses

Now we are using the Link Checker API the values being persisted in the
`status` attribute are different, so we update URLStatusPresentation to handle
these.  Also we now have errors and warnings sent from Link Checker API, so we
display these alerts on the Link edit page.

emmabeynon committed:
Change presentation of link statuses

Now we are using the Link Checker API the values being persisted in the
`status` attribute are different, so we update URLStatusPresentation to handle
these.  Also we now have errors and warnings sent from Link Checker API, so we
display these alerts on the Link edit page.

kevindew commented:
@thomasleese you might want to cherry pick out the autoloading commit and put that in as a separate PR so that can get merged early and the benchmarking team can use it - seems like it's a likely source of merge conflicts.

emmabeynon committed:
Change default value for link_errors and link_warnings

Link Checker API returns an empty hash for link errors and link warnings, so
the default value for `link_errors` and `link_warnings` in Links and
LocalAuthorities should also be an empty hash.  We also update tests to reflect
this.

emmabeynon committed:
Update BadLinksURLAndStatusExporter CSV

Now that this app is using Link Checker API to check it links, more detailed
information about the results of the link check are now available via
`link_errors` and `link_warnings`.  The BadLinksUrlAndStatusExporter was
previously reporting on a link's `status`, which will now only return "ok",
"broken" or "caution".  This isn't enough information in order to take an
appropriate action on the link, so we now include `link_errors` and
`link_warnings` in the CSV export.

sihugh commented on lib/tasks/check-links/link_checker.rake:
```diff
@@ -1,31 +1,20 @@
-      begin
-        Rails.logger.info("Lock obtained, starting link checker")
-        Services.icinga_check(service_desc, true, "Lock obtained, starting link checker")
+  begin
```
What was the reasoning behind removing the distributed lock?  We still only want to run this on one machine don't we?

thomasleese replied:
Hmm, perhaps I misunderstood what the point of the lock was, I figured it was there to avoid someone accidentally running the take command twice. Since the rake task now only takes a few minutes (it sends off requests to the `link-checker-api` which then responds through a web hook) is it necessary to have the lock?

sihugh replied:
It's less the amount of time the request takes, rather that we probably don't want to update all the links multiple times if we can avoid it.

thomasleese replied:
That shouldn't be a problem. The Link Checker API automatically caches results for 24 hours for us anyway, and we only need to perform the update queries if the last checked time has been updated.

thomasleese committed:
Minor code improvements

emmabeynon committed:
Update Link model callbacks

Now that this app is using the Link Checker API, more link check information is
being persisted in the new `link_errors` and `link_warnings` attributes.  This
updates the callbacks on the Link model so that these attributes are set and
updated appropriately when a link is created or updated.  We also take the
opportunity to rename some methods names that weren't very clear.

thomasleese committed:
Add support for secure webhooks

kevindew commented on app/controllers/webhooks_controller.rb:
```diff
@@ -0,0 +1,30 @@
+      )
+    )
+  end
+
```
Should probably have these as private methods

kevindew commented on app/controllers/webhooks_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  end
+
+  def verify_signature
+    given_signature = request.env["HTTP_X_LINKCHECKERAPI_SIGNATURE"]
```
can we use a request.headers method? looking at [source](https://github.com/rails/rails/blob/d2dde6cbfc11bf39be30d22e2ee7c41c119e0d51/actionpack/lib/action_dispatch/http/request.rb#L188) they header might be formatted in header spec there (or maybe it's a nightmare because they format them different)

kevindew commented on app/controllers/webhooks_controller.rb:
```diff
@@ -0,0 +1,30 @@
+    return head :bad_request unless given_signature
+
+    request.body.rewind
+    body = request.body.read
```
rather than needing to do this can we just use `request.raw_post` ?

kevindew commented on app/models/link.rb:
```diff
@@ -56,31 +58,35 @@ def self.build(params)
-  def existing_link_url
-    @_link ||= Link.where(url: self.url).distinct.first
+  def existing_link
+    @_link ||= Link.where(url: self.url).group_by(&:url)[self.url].try(:first)
```
Do we gain anything here over doing `Link.where(url: self.url).first`?

emmabeynon replied:
No, I think you're right.  Will change!

kevindew commented on app/presenters/url_status_presentation.rb:
```diff
@@ -3,17 +3,26 @@ module UrlStatusPresentation
+      "Broken: #{link_errors.keys.first}"
+    elsif status == "caution"
+      "Note: #{link_warnings.keys.first}"
+    else
```
Might we be dealing with pending links?

emmabeynon replied:
I don't think we'll be sending back any pending links...am I correct in thinking that, @thomasleese?

kevindew replied:
Sure, I guess if your only source is webhook and that's not firing until complete you won't get any pendings

thomasleese replied:
In my head, I had thought we were ignoring links when they came back as pending, but I'm no longer sure that's the case, would be worth checking.

thomasleese replied:
Yes, what @kevindew said makes more sense than what I did!

kevindew commented on lib/local-links-manager/check_links/link_status_requester.rb:
```diff
@@ -0,0 +1,42 @@
+module LocalLinksManager
+  module CheckLinks
+    class LinkStatusRequester
+      delegate :url_helpers, to: "Rails.application.routes"
```
🎉 🌮

kevindew commented on lib/local-links-manager/check_links/link_status_updater.rb:
```diff
@@ -1,40 +1,43 @@
+        LocalAuthority
+          .where(homepage_url: link_report.uri)
+          .link_last_checked_before(link_report.checked)
+          .update_all(**fields)
```
Do these need the double splat operator? would have thought they had no effect

emmabeynon replied:
Tests pass without the double splat operator, so will remove them ⚡️

kevindew reviewed (commented):
all looking good 👍

thomasleese committed:
Add migrations to support new link checker api messages

thomasleese committed:
Update code to support new messages format

thomasleese committed:
Fix secure webhook controller

thomasleese committed:
Output suggested fix, errors and warning correctly

kevindew commented on lib/local-links-manager/export/bad_links_url_and_status_exporter.rb:
```diff
@@ -3,22 +3,22 @@
-          LocalAuthority.where.not(status: "200").distinct.pluck(:homepage_url, :status).each do |la|
-            csv << la
+          LocalAuthority.where.not(status: "ok").pluck(:homepage_url, :link_errors, :link_warnings).each do |la|
+            csv << [la[0], la[1].join(','), la[2].join(',')]
```
not sure about the wisdom of using comma separators here since we have commas in some of the messages. I wonder if we could just have status and problem summary and not worry about the lists?

thomasleese committed:
Update bad links exporter to include problem summary

thomasleese committed:
Restore the distributed lock when calling check-links

thomasleese committed:
Add an inclusion list for valid link statuses

Also update the link status exported to use problem summary instead
of status.

thomasleese committed:
Update link status exporter to include status column

kevindew reviewed (approved):
I've not found anything else to nag about so I'm going to say LGTM 👍  😀  - probably worth getting a look from someone on the team that owns this

brenetic commented:
🚀  Looks good!! 🚀
